donejs -> donejs-cli (for npm link)

leoj3n opened the pull request:
Related to https://github.com/donejs/cli/pull/83

Once these two PRs are merged, it will be possible to do:

```
git clone -b rename-binary-for-npm-link-version --single-branch https://github.com/donejs/donejs.git && cd donejs && npm install && npm link

cd .. && git clone -b rename-binary-for-npm-link --single-branch  https://github.com/donejs/cli.git && cd cli && npm install && npm link

cd .. && mkdir fakeapp && cd fakeapp && npm link donejs-cli && donejs add app
```

If you try and do this without these PRs, `donejs/donejs` and `donejs/cli` will collide upon `npm link`, because they have the same binary file name of "donejs".

These PRs make the `donejs` binary be "donejs" and the `donejs-cli` binary be "donejs-cli".

So, these PRs enable local development by use of `npm link`, instead of having to [hack](https://github.com/donejs/developer-relations/issues/29) the [files](https://github.com/donejs/cli/pull/82#issuecomment-289615878).

leoj3n committed:
donejs -> donejs-cli (for npm link)

phillipskevin commented on lib/cli/cmd-init.js:
```diff
@@ -19,10 +19,14 @@ module.exports = function(folder, opts) {
-        .then(function() {
-          return runCliInit(folderPath, opts);
-        });
+      if (!fs.existsSync(path.join(folderModules, 'donejs-cli'))) {
```
Is there a reason to not overwrite the cli bin if it already exists? If there is, you'll probably need to write a test for this case.

leoj3n replied:
Yes, there is a reason: Essentially, if you let `installCli` run after having done `npm link`, that will for some reason remove some of the deps from the linked `donejs/cli`'s `node_modules`, such as `q`. I'm still not entirely sure why the `npm install` command from `installCli` causes this to happen in the linked package, but turned out not to be something to have to worry about since I figured I could add this if check. Was planning on adding tests for this today, but hadn't really thought about what tests were needed yet, so thanks for the suggestion here.

leoj3n replied:
I think I've added the necessary tests.

matthewp replied:
Can you link to the test you added for this?

leoj3n replied:
https://github.com/donejs/donejs/blob/f3187f7874ad981f19a55df78493c15cadda4f8a/test/cmd-init-test.js#L100-L116

phillipskevin commented on lib/cli/run-binary.js:
```diff
@@ -8,11 +8,17 @@ module.exports = function(args, options, types) {
+      // backwards compatiblilty for bin/donejs before it became bin/donejs-cli
       if (!fs.existsSync(donejsBinary)) {
-        var msg = 'Could not find local DoneJS binary (' + donejsBinary + ')';
+        doneScript = process.platform === 'win32' ? 'donejs.cmd' : 'donejs';
```
You should add a test for this.

leoj3n replied:
Thanks, will do!

matthewp replied:
Was this test added?

leoj3n replied:
Yes, this test proves the backwards compatibility: https://github.com/donejs/donejs/blob/f3187f7874ad981f19a55df78493c15cadda4f8a/test/run-binary-test.js#L70-L98

phillipskevin commented:
@matthewp do you have any concerns about this? I'm not sure if this would be considered a breaking change.

matthewp reviewed (approved):
👍 LGTM

leoj3n committed:
CONTRIBUTING info regarding donejs-cli

leoj3n committed:
Test that runBinary is able to locate donejs-cli

leoj3n committed:
Test that install skips if node_modules/donejs-cli

Install command will skip installing donejs-cli if it already
exists, assuming it was put there by `npm link` or otherwise
for debugging and/or development purposes.

leoj3n committed:
CONTRIBUTING note about npm link generator-donejs

matthewp reviewed (changes requested):
I'm worried that skipping installing donejs-cli could be a source of bugs. The expectation when you run `donejs add app` in an existing folder is that it will init a donejs app within that folder. This change prioritizes a fairly rare development need over that scenario.

I think a better approach to enable this behavior is with a flag. We already have a `--skip-install` flag so it wouldn't be weird to add another. I'd suggest `--debug`, we could add other capabilities behind this flag in the future.

leoj3n commented:
@matthewp Some thoughts (correct me if I'm wrong):

A.) People won't run `donejs add app` more than once within the same directory. I imagine that's not supported, similar to how running `donejs add plugin` inside a generated app directory is not.

B.) People will not somehow already have a `node_modules/donejs-cli` in the otherwise empty directory that they are running `donejs add app` outside of the debug/development case.

C.) I think using `npm link` is common knowledge, and most people would expect to be able to do that without having to pass a debug flag. Similar to how `npm link` already works w/ the generators.

D.) A debug flag would just be paving over the issue where `npm install` (or whatever command is getting run in the linked `donejs-cli`) [causes deps to disappear](https://github.com/donejs/donejs/pull/892#discussion_r109319447). I haven't investigated this yet.

E.) I added the [`if statement`](https://github.com/donejs/donejs/blob/f3187f7874ad981f19a55df78493c15cadda4f8a/lib/cli/cmd-init.js#L22) to skip running `installCli` regardless of that strange behavior above (where letting `installCli` run while `donejs-cli` was pre-linked caused certain deps to disappear out of the linked `donejs-cli`'s `node_modules`). I feel that `if` makes sense there regardless, because obviously `donejs-cli` exists, so no need doing the work of `installCli`.

F.) The only case I can imagine where skipping `installCli` if `node_modules/donejs-cli` already exists might be a problem is when trying to run `donejs add app` in an already generated app, which is not supported -- the whole point of installing `donejs-cli` to the generated app itself is to lock down the version of `donejs-cli` to the version of `donejs` used to generate the app.

---

Maybe I missed something, and `donejs add app` is a way of upgrading an app? [Related](https://github.com/donejs/donejs/blob/master/docs/guides/migrate-1.md).

Running `donejs add app` a second time asks for "Project main folder", "Description", etc and overwrites `package.json`, so I would expect my assumption here is correct, and any update functionality would be implemented as a new separate command that is not affected by that `if`.

Running `donejs add app` in a clean directory with `npm i -g donejs@0.9.2`, committing changes, and then running `donejs add app` again with `npm i -g donejs@1` results in this diff:

```diff
diff --git a/package.json b/package.json
index c948964..08716c6 100644
--- a/package.json
+++ b/package.json
@@ -33,25 +33,30 @@
     "npmAlgorithm": "flat"
   },
   "dependencies": {
-    "can": "^2.3.16",
-    "can-connect": "^0.5.0",
-    "can-zone": "^0.5.0",
-    "done-serve": "^0.2.0",
-    "done-autorender": "^0.8.0",
-    "done-component": "^0.4.0",
-    "done-css": "~2.0.2",
-    "generator-donejs": "^0.9.0",
-    "jquery": "~2.2.1",
-    "steal": "^0.16.0"
+    "can-component": "^3.0.7",
+    "can-connect": "^1.3.8",
+    "can-define": "^1.0.17",
+    "can-route": "^3.0.8",
+    "can-route-pushstate": "^3.0.3",
+    "can-set": "^1.1.0",
+    "can-stache": "^3.0.20",
+    "can-view-autorender": "^3.0.4",
+    "can-zone": "^0.6.6",
+    "done-serve": "^1.0.0",
+    "done-autorender": "^1.0.0",
+    "done-component": "^1.0.0",
+    "done-css": "^3.0.0",
+    "generator-donejs": "^1.0.0",
+    "steal": "^1.3.1",
+    "steal-less": "^1.2.0",
+    "steal-stache": "^3.0.5"
   },
   "devDependencies": {
-    "documentjs": "^0.4.2",
-    "donejs-deploy": "^0.4.0",
-    "funcunit": "~3.0.0",
-    "steal-qunit": "^0.1.1",
-    "steal-tools": "^0.16.0",
-    "testee": "^0.2.4",
-    "donejs-cli": "^0.9.4",
-    "can-fixture": "^0.1.2"
+    "donejs-cli": "^1.0.0",
+    "funcunit": "^3.2.0",
+    "steal-qunit": "^1.0.1",
+    "steal-tools": "^1.2.0",
+    "testee": "^0.3.0",
+    "can-fixture": "^1.0.13"
   }
 }
```

So, running `donejs add app` a second time in an already generated app directory does cause the `donejs-cli` in `package.json` to be modified when using a different version of `donejs`, although it does not look like this is intended to be a supported use case for the `donejs add app` command.

matthewp commented:
I definitely would expect to be able to run `donejs add app` in an already generated app, or a partially generated app. It would be bizarre to me if it were to behave differently depending on whether a node_modules/donejs-cli folder already exist (but ignore all other folders).

Given that this PR is specifically addressing contributors to donejs and not donejs users, I think a small burden of adding a flag is warranted over the potential (however unlikely) of a donejs user getting unexpected output from running `donejs add`.

An alternative I think is acceptable would be if it checked if the existing donejs-cli folder is a symlink (which I think it would be with `npm link`) rather than just checking if it exists. A symlink would express an intent.

leoj3n commented:
@matthewp, @phillipskevin 

The reason `donejs add app` without this PR _removes_ installed `node_modules` in the linked `donejs-cli` (on [donejs/cli@rename-binary-for-npm-link](https://github.com/donejs/cli/pull/83)) is because npm@3 uses a maximized, flat-tree.

Essentially, when `donejs add app` triggers `installCli`, that runs `npm install donejs-cli@1.0.x --loglevel error` in the new app directory, and if `node_modules/donejs-cli` already exists there (via `npm link`), it removes the `donejs-cli` dependencies from `node_modules/donejs-cli/node_modules` and it "flattens" them into the project's `node_modules`.

Then, because `npm link` creates a symlink, the linked `donejs-cli` cannot find it's dependencies, because when it searches upwards in the filesystem, it cannot find those deps relative to the actual location of `donejs-cli`.

After getting "flattened", `npm outdated` in the linked `donejs-cli` is:

```
Package             Current  Wanted  Latest  Location
commander           MISSING   2.9.0   2.9.0  donejs-cli
concat-stream       MISSING   1.6.0   1.6.0  donejs-cli
cross-spawn-async   MISSING   2.2.5   2.2.5  donejs-cli
debug               MISSING   2.6.3   2.6.3  donejs-cli
mkdirp              MISSING   0.5.1   0.5.1  donejs-cli
q                   MISSING   1.5.0   1.5.0  donejs-cli
yeoman-environment  MISSING   1.6.6   1.6.6  donejs-cli
mocha                 2.5.3   2.5.3   3.2.0  donejs-cli
mockery               1.7.0   1.7.0   2.0.0  donejs-cli
```

The linked `donejs-cli` will not be able to find it's dependencies such as `q` because it searches upwards relative to the actual location, not the symlinked location:

```
module.js:471
    throw err;
    ^

Error: Cannot find module 'q'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/Users/leoj/bitovi/,=/donejs_donejs_issue-892/donejs-cli/lib/utils.js:1:71)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)

Error: Command `/Users/leoj/bitovi/,=/donejs_donejs_issue-892/ftestapp/node_modules/.bin/donejs-cli` did not complete successfully
    at ChildProcess.<anonymous> (/Users/leoj/bitovi/,=/donejs_donejs_issue-892/donejs/lib/utils.js:51:23)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)
```

It seems like this would be a somewhat common problem with `npm link` and flattening, and it appears it is: https://github.com/npm/npm/issues/10343

@matthewp It doesn't look like that npm@3 issue will be resolved any time soon. Can you think of any solution that still lets `installCli` run with these new facts explained? That would be the best.

Otherwise, I agree the checking if symlink is our best bet as long as npm@3 flattens linked deps.

leoj3n committed:
Skip installing donejs-cli only if symlink exists

frank-dspeed commented:
@chasenlehara importent info for you 

# 3 ways for local dev
1. using docker the most isolated and best way allows also multi nodejs version tests msg me about instructions if needed
2. install local git repo via npm install git+/path/to/dev
3. ln -s /path/to/dev ./node_modules/donejs-cli

greetings :)

frank-dspeed commented:
I think a more good thing would be to discusse replacing donejs with donejs-generator nativ yo and let that handle versions. and all that via a canjs/canjs/package.json i think there is no need to have a backward compatible donejs binary as its only external tooling and has nothing to do with the project it self
only conflict that can happen is stealjs 0.x -> 1.x and we can handle that via codemods

https://github.com/donejs/cli/issues/63
